### ISSUE TITLE
Continuous reindexing slows download: allow reindexing to be delayed until after all downloads are complete

### ISSUE THREAD

**Issue by Tillomar** [osmandapp/OsmAnd#14893]
**Problem:**
Reindexing takes a lot of time, especially if you have a big map base like me (>300 GB) which is also located on SDCard.

Updating a map base will get worse with size on two accounts -- first, there are more files to update, therefore there are more reindexing runs inbetween; -- and second, each reindexing run will take longer as the map base gets bigger. Essentially, we are talking O(n^2).

Also, OSMAND's GUI is nearly unaccessable during reindexing due to processing load. (This may be dependant on the device.)

**Proposed solution:**
Delay the reindexing until a better time -- for example, until after all pending downloads have completed.

Another option would be to start reindexing when the app is being closed, effectivly delaying the termination of the app until reindexing is done. This could be combined with a request to the user like
`"There is some new map data not currently integrated into your map pool. Would you like to integrate it now? This would be done in the background before closing the APP."`
which will pop up again next time if the user chose to deny the reindexing operation.

There are a lot of other possible solutions/options, which may also be configurable to the user -- but you get the gist: DO NOT reindex for each modified file.

**Comment by vshcherb** [comment]
Reindexing only reindexes new maps, it probably checks all maps but doesn't read its structure again.

**Comment by scaidermern** [comment]
It sounds like a good idea to postpone reindexing until the last map download has been completed (or aborted).

**Comment by Tillomar** [comment]
Will anyone be doing anything about that?
Just asking... 😊

**Comment by vshcherb** [comment]
We don't know what causes 60 sec reindexing and we can't reproduce it. Delaying process will still cause same 60sec in total but the issue it shouldn't take that long

**Comment by dreirund** [comment]
> We don't know what causes 60 sec reindexing

Device's processing power or/and memory card data read speed and total amount of maps installed?

I am on device codename 'markw', which by now can be counted as "pretty old".

> Delaying process will still cause same 60sec in total but the issue it shouldn't take that long

It will do the 60sec, 120sec, or whatever the amount is on the individual local machine only _once_, not after each single download.

**Comment by Tillomar** [comment]
There are two points here:

1. The context which might cause the reindex being slow: processing power, memory access timing, total amount of map memory/map file count, other tasks running in parallel, ...;
2. my observation that a reindex is done any time a single file download has finished is contested.

For 1.) I agree that my old machine – it is a Nokia 5.1 which was middle class at best when it came into the market, and the SDCard interface was especially slow – might have been part of the problem. AND I had nearly the complete world map including wikipedia, hillshapes and contour lines on the SD card, even if that was a fast one. But anyway, even if reindexing takes some time, it would not matter much if it is done once.

For 2.) OSMAND itself displays a message that it is indexing the maps, together with the current map being indexed. And that name iterates through all map files (revers alphabetical order, actually). When indexing is done, OSMAND will download the next file, and then AGAIN index ALL files. – I just checked this behavior again, and also made a video off it if anyone needs proof.

Therefore my request for change stays as it is, because the problem is not the total amount of time one reindex run takes, but the repetition of that reindexing causing so much extra delay.

The version of OSMAND on that old machine is 4.1.11.
On my new one, which is a Pixel 9a, I don't have a SD card, nor enough memory to hold all map files. Currently only the maps from Central Europe are configured. On the other hand, that machine is much faster, and also the only memory which is internal is much faster. When I update multiple map files, there is also something shown after every file, but it's displayed too short to see if it is reindexing multiple or all files. – The OSMAND version on the Pixel is 5.0.5.

**Comment by dreirund** [comment]
> *For 2.) OSMAND itself displays a message that it is indexing the maps, together with the current map being indexed. And that name iterates through all map files (revers alphabetical order, actually). When indexing is done, OSMAND will download the next file, and then AGAIN index ALL files. – I just checked this behavior again, and also made a video off it if anyone needs proof.*

I exactly observe the same.

I use the newest version of OsmAnd~ available -- so up to date.

**Comment by jlariviere1** [comment]
I have the most current version of OsmAnd installed (through F-Droid).  
I updated one 20 MB map (I have a Lot of map updates to install, many GB worth).  It took about 2 minutes to index / check the index files with that one 20 MB file.  The last time I tried to update, it made the phone unusable and had to be rebooted several times and took many hours, and was way beyond the time it would take to download and index once.  

It would be extremely helpful if it didn't take 2-5 minutes after every download and waited to check until after the updates were finished.  I really hope this is being considered!  I do financially support this and other projects, encourage others to do so, and appreciate the ongoing development work.  This would definitely help those who travel or need a larger map area downloaded and need to have more than just a single map installed.

**Comment by SKBotNL** [comment]
I also have this issue when the live updates are being downloaded, I have live updates enabled for quite a lot of regions, which causes OsmAnd to become completely unusable for about 1-2 minutes.

Is it possible to run the downloading and indexing off of the UI thread? It would at least make the app usable during updates.

**Comment by mazay** [comment]
There are a bunch of duplicates of this issue, and this one is 4 years old - it's very unfortunate that it's not being looked into.

I have an older/simpler android phone for motorcycle navigation, with Europe maps (standard, contour, terrain); the total size of the maps is over 80 GB. The device doesn't have the storage for that, so I'm using an SD card. It takes a couple of days to update all of them; adding new maps also takes a significant amount of time.

I would imagine this issue is not noticeable on newer devices with enough local storage or fewer maps; however, as stated many times before, the issue exists and is a serious usability problem.

Thus, I'm asking developers to pay attention to this issue - the free version of the app with limited map downloads may not be affected as much, but paying customers with the limit unlocked are really suffering.

**Comment by mazay** [comment]
Ok, I just found this in the official docs - https://osmand.net/docs/user/troubleshooting/maps-data#maps-slowly-loading-on-android-11-12-sd-card

Followed option 2, and indexing is much faster now.

**Comment by dreirund** [comment]
> Ok, I just found this in the official docs -
> https://osmand.net/docs/user/troubleshooting/maps-data#maps-slowly-loading-on-android-11-12-sd-card
> 
> Followed option 2, and indexing is much faster now.

This is introduced to be of a concern only for Android 11 onwards.  
I have it with Android 7 already, and for me it very much seems like
CPU time bound on the slow device.

So indexing only _after_ having completed (or aborted) all ongoing
downloads _still_ is a very valid request.
